File: DesktopApp/loadFromJson.py
import mod, json, threading, os

# Convert json data for a mod into a mod object
def _dictToMod(data):
    priorityData = data["priority"]
    priorityLevel = mod.Priority(priorityData["name"], red=priorityData["r"], green=priorityData["g"], blue=priorityData["b"])

    if "tablePosition" in data:
        tablePos = data["tablePosition"]
    else:
        tablePos = -1
        
    modObj = mod.Mod(modName=data["name"], modPriority=priorityLevel, modVersions=data["versions"], url=data["url"], modID=data["id"], tablePosition=tablePos)
    return modObj

# Convert json data for a mod profile into a mod profile object. Create each mod object on its own thread, as each will make an API call.
def _dictToModProfile(data, requireValidModURL=True):
    threadList = []
    
    newModList = []
    modListData = data["modlist"]

    def threadHelper(newMod, list, requireValidURL=True):
        modObj = _dictToMod(newMod)

        if not requireValidURL or modObj.isValid():
            list.append(modObj)

    for newMod in modListData:
        thread = threading.Thread(target=threadHelper, args=(newMod, newModList, requireValidModURL))
        threadList.append(thread)
        thread.start()

    for thread in threadList:
        thread.join()

    return mod.Profile(modList = newModList, name=data["name"], selectedVersion=data["version"])

# Open a json file, read the data, and create a list of mod profiles from it
def createProfileList(filename="mods.json"):
    appdata = os.getenv('APPDATA')
    directory = os.path.join(appdata, 'ModTracker')
    os.makedirs(directory, exist_ok=True)
    json_path = os.path.join(directory, filename)

    newProfileList = []
    try:
        with open(json_path, "r") as f:
            data = json.load(f)
            for entry in data:
                newProfileList.append(_dictToModProfile(entry, requireValidModURL=False))
    except FileNotFoundError or json.JSONDecodeError or json.decoder.JSONDecodeError:
        print("Failed to read file. No data recovered.")
        return []
    return newProfileList

# Open a json file, read the data, and create a single mod profile from it
def createProfile(filename="mods.json", rawJson = None, requireValidModURL=True):
    if rawJson:
        profile = _dictToModProfile(rawJson, requireValidModURL=requireValidModURL)
        return profile

    appdata = os.getenv('APPDATA')
    directory = os.path.join(appdata, 'ModTracker')
    os.makedirs(directory, exist_ok=True)
    json_path = os.path.join(directory, filename)

    try:
        with open(json_path, "r") as f:
            data = json.load(f)
            profile = _dictToModProfile(data, requireValidModURL=requireValidModURL)
    except FileNotFoundError or json.JSONDecodeError or json.decoder.JSONDecodeError:
        print(f"Failed to read file at {json_path}. No profile created.")
        return None
    return profile

File: WebApp/src/data/loadFromJson.js
import mod from './mod.js'
import profile from './profile.js'

// convert JSON data into a priority object
function createPriority(data) {
    if (data !== undefined && Object.keys(data).length === 0) {
        let error = new TypeError("Cannot load priority from this data. Dictionary is blank.")
        error.name = "Blank priority data"
        throw error
    }

    return new mod.Priority(
        data?.name,
        data?.r,
        data?.g,
        data?.b
    )
}

// create a list of priority objects from JSON data
function createPriorityList(dataList) {
    const newPriorityList = []

    for (const entry of dataList) {
        const priorityObj = createPriority(entry)
        newPriorityList.push(priorityObj)
    }

    return newPriorityList
}

// convert JSON data into a mod object
function createMod(data) {
    if (data === undefined) {
        let error = new TypeError("Cannot load mod from this data. Data is undefined.")
        error.name = "Undefined mod data"
        throw error
    } else if (Object.keys(data).length === 0) {
        let error = new TypeError("Cannot load mod from this data. Dictionary is blank.")
        error.name = "Blank mod data"
        throw error
    }

    const priorityLevel = createPriority(data.priority)

    return new mod.Mod(
        data.url,
        priorityLevel,
        data.tablePosition,
        data.name,
        data.id,
        data.versions
    )
}

// convert JSON data into a profile object
function createProfile(data, requireValidModURL = true) {
    if (data === undefined) {
        let error = new TypeError("Cannot load profile from this data. Data is undefined.")
        error.name = "Undefined profile data"
        throw error
    } else if (data.modlist === undefined) {
        let error = new TypeError("Cannot load profile from this data. Data.modlist is undefined.")
        error.name = "Undefined profile modlist"
        throw error
    }

    const newModList = []

    for (const entry of data.modlist) {
        const modObj = createMod(entry)

        if (!requireValidModURL || modObj.verifyURL()) {
            newModList.push(modObj)
        }
    }

    return new profile.Profile(newModList, data.version, data.name)
}

// create a list of profile objects fron JSON data
function createProfileList(dataList, requireValidModURL = true) {
    const newProfileList = []

    for (const entry of dataList) {
        const profile = createProfile(entry, requireValidModURL)
        newProfileList.push(profile)
    }

    return newProfileList
}

// convert JSON data into a profile manager object
function createProfileManager(data) {
    if (data === undefined) {
        let error = new TypeError("Cannot load profile manager from this data. Data is undefined.")
        error.name = "Undefined manager data"
        throw error
    } else if (Object.keys(data).length === 0) {
        let error = new TypeError("Cannot load profile manager from this data. Dictionary is blank.")
        error.name = "Blank manager data"
        throw error
    }

    const profileList = createProfileList(data.profileList)
    const priorityList = createPriorityList(data.priorityList)

    return new profile.ProfileManager(profileList, priorityList)
}

export default {
    createMod,
    createProfile,
    createProfileManager
}
